### PR TITLE
Use decorated names for linked_symbols on Windows

### LINKED ISSUE #96423: unresolved external symbol _DllMain
### Code

I tried this code:

```toml
[lib]
crate_type = ["cdylib"]
```
```rust
#[no_mangle]
extern "system" fn DllMain(module: HINSTANCE, reason: u32, _reserved: *mut c_void) -> i32 {
    if reason == DLL_PROCESS_ATTACH {
        unsafe {
            DisableThreadLibraryCalls(module);
        }

        thread::spawn(move || unsafe { dll_attach() });

        return 1;
    }

    return 0;
}
```

```
error: linking with `link.exe` failed: exit code: 1120
```
```
symbols.o : error LNK2001: unresolved external symbol _DllMain
  Hint on symbols that are defined and could potentially match:
    _DllMain@12
C:\Users\random\Desktop\project\target\i686-pc-windows-msvc\debug\deps\test.dll : fatal error LNK1120: 1 unresolved externals
```

It seems like `#[no_mangle]` is not working anymore.

### Version it worked on

`rustc --version --verbose`:
```
rustc 1.62.0-nightly (18f314e70 2022-04-24)
binary: rustc
commit-hash: 18f314e7027fe7084aaab8620c624a0d7bd29e70
commit-date: 2022-04-24
host: x86_64-pc-windows-msvc
release: 1.62.0-nightly
LLVM version: 14.0.1
```

### Version with regression

`rustc --version --verbose`:
```
rustc 1.62.0-nightly (055bf4ccd 2022-04-25)
binary: rustc
commit-hash: 055bf4ccd521c2c2185166c86951be7be145727c
commit-date: 2022-04-25
host: x86_64-pc-windows-msvc
release: 1.62.0-nightly
LLVM version: 14.0.1
```
</p>
</details>

### PR DESCRIPTION
**Pull request by nbdd0121**
Fix #96423 

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/back/symbol_export.rs [hunk 1]
```diff
@@ -493,6 +493,76 @@ pub fn symbol_name_for_instance_in_crate<'tcx>(
...
+    let target = &tcx.sess.target;
+    if !target.is_like_windows {
+        // Mach-O has a global "_" suffix and `object` crate will handle it.
+        // ELF does not have any symbol decorations.
+        return undecorated;
+    }
+
+    let x86 = match &target.arch[..] {
+        "x86" => true,
+        "x86_64" => false,
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/symbol_export.rs:517]: Looks like this case also always results in undecorated names.

File: compiler/rustc_codegen_ssa/src/back/symbol_export.rs [hunk 2]
```diff
@@ -493,6 +493,76 @@ pub fn symbol_name_for_instance_in_crate<'tcx>(
...
+            }
+            return undecorated;
+        }
+    };
+
+    let args_in_bytes: u64 = args
+        .iter()
+        .map(|abi| abi.layout.size.bytes().next_multiple_of(target.pointer_width as u64 / 8))
+        .sum();
+    format!("{prefix}{undecorated}{suffix}{args_in_bytes}")
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/symbol_export.rs:563]: I've seen this logic elsewhere, it's split into `fn i686_decorated_name` and `fn i686_arg_list_size` as a part of implementation of the raw-dylib feature.

Could you make sure these two implementations do the same thing?
`fn i686_decorated_name`, for example, adds or not adds underscore depending on the `mingw` flag.
(Ideally the code would be shared, but not necessarily in this PR.)

File: src/test/ui/symbol-names/x86-stdcall.rs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+// check-pass
```
Comment (petrochenkov) [inline src/test/ui/symbol-names/x86-stdcall.rs:1]:
```suggestion
// build-pass
```
`check-pass` won't run codegen and linking stages.

File: compiler/rustc_codegen_ssa/src/back/symbol_export.rs (hunk 1 above)
Comment (nbdd0121) [reply compiler/rustc_codegen_ssa/src/back/symbol_export.rs:517]: x86_64 code can still use "vectorcall" convention

Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/symbol_export.rs:517]: The linked MS doc says "Note that in a 64-bit environment, functions are not decorated.", apparently it applies to vectorcall too?
Raw-dylib didn't attempt to support vector call yet, so I don't know what is the correct behavior there.

Comment (nbdd0121) [reply compiler/rustc_codegen_ssa/src/back/symbol_export.rs:517]: That's an inaccurate statement, it should rather say that functions are not decorated *by default*. Vector call is still a separate calling convention on x64 however, and is still decorated (https://docs.microsoft.com/en-us/cpp/cpp/vectorcall?view=msvc-170, https://godbolt.org/z/asdbdzePz)

File: compiler/rustc_codegen_ssa/src/back/symbol_export.rs (hunk 2 above)
Comment (nbdd0121) [reply compiler/rustc_codegen_ssa/src/back/symbol_export.rs:563]: I just checked that the current behaviour in i686_decorated_name to not add `_` in MinGW is due to the fact that it uses dlltool to create a library, and i686 dlltool will automatically add `_` to symbols if it does not start with `@`. For the generated object files we need to add `_` ourselves.

There are no other behavioural differences.

File: compiler/rustc_codegen_ssa/src/back/symbol_export.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/symbol_export.rs:517]: >Vector call is still a separate calling convention on x64

Yeah, I was aware of that, but wasn't sure about the decoration.
Just checked it with Visual Studio's C compiler, and looks like it's indeed the only calling convention that is decorated on x64.

Comment (nbdd0121) [reply compiler/rustc_codegen_ssa/src/back/symbol_export.rs:517]: Filed MicrosoftDocs/cpp-docs#3854

Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/symbol_export.rs:517]: Ha, I didn't even know this stuff is in open source now.

**Pull request merged**